src/beignet/_newton_raphson.py

```diff
@@ -0,0 +1,172 @@
+
+# Define function signature types
+FuncType = Callable[[Tensor, Tensor], tuple[Tensor, Any]]
+FPrimeType = Callable[[Tensor, Tensor, Any], Tensor]
+
+
+def newton_raphson(
+    func: FuncType,
+    f_prime: FPrimeType,
+    params_batch: Tensor,
+    y_target_batch: Tensor,
+    x_init: Union[float, Tensor],
+    max_iter: int = 50,
+    tol: float = 1e-5,
+    min_bound: Optional[float] = None,
+    max_bound: Optional[float] = None,
+    max_update_step: float = 1.0,
+) -> Tensor:
+    """
+    Performs batched Newton-Raphson root finding: finds x such that
+    func(x, params) = y_target. # <-- Wrapped line 26
+
+    Uses update step clamping for stability.
+
+    Args:
+        func: A function `f(x_batch, params_batch)` -> (y_batch, aux_data).
+        f_prime: A function `f'(x_batch, params_batch, aux_data)` -> y_prime_batch.
+        params_batch: Batch of parameters specific to each function instance. (B, ...).
+        y_target_batch: Target values for each instance (B,).
+        x_init: Initial guess for x. Scalar or Tensor (B,).
+        max_iter: Maximum number of iterations.
+        tol: Absolute tolerance for convergence check (|y_val - y_target| < tol).
+        min_bound: Lower bound to clamp x during iteration.
+        max_bound: Upper bound to clamp x during iteration.
+        max_update_step: Maximum absolute value allowed for the x update step.
+                         Must be positive. # <-- Wrapped line 40
+
+    Returns:
+        The found roots x_batch (B,). Non-converged elements have last computed value.
+    """
+    B: int = params_batch.shape[0]
+    device: torch.device = params_batch.device
+    dtype: torch.dtype = params_batch.dtype  # Use dtype from params for consistency
+
+    if not torch.is_tensor(y_target_batch):
+        y_target_batch = torch.tensor(y_target_batch, device=device, dtype=dtype)
+    if y_target_batch.ndim == 0:  # If single target, expand to batch
+        y_target_batch = y_target_batch.expand(B)
+    if y_target_batch.shape[0] != B:
+        raise ValueError("y_target_batch must have shape (B,) or be a scalar.")
+
+    if max_update_step <= 0:
+        raise ValueError("max_update_step must be positive.")
+
+    # Initialize x
+    x: Tensor
+    if isinstance(x_init, torch.Tensor):
+        x = x_init.clone().to(device, dtype)
+        if x.shape[0] != B:
+            raise ValueError(
+                f"x_init tensor shape {x.shape} incompatible with batch size {B}"
+            )
+    else:
+        # Ensure x is created with the correct dtype
+        x = torch.full_like(y_target_batch, float(x_init), dtype=dtype)
+
+    # Ensure initial x respects bounds
+    if min_bound is not None:
+        x = torch.clamp(x, min=min_bound)
+    if max_bound is not None:
+        x = torch.clamp(x, max=max_bound)
+
+    # Track active elements (those not yet converged)
+    active: Tensor = torch.ones_like(y_target_batch, dtype=torch.bool)
+
+    for i in range(max_iter):
```

Comment:
You need to add monotonicity and oscillation guards or it’ll ping pong between sign changes.

Reply:
yeah you can see it happen in the test case if you increase the `max_update_step` value